app/core/store/modules/unitMap.js

```diff
@@ -0,0 +1,84 @@
+      1. levels list from the classroom snapshot
+      2. position, nextLevels, first from the classroom snapshot, if does not exist then from campaign snapshot
+      4. any other data from the campaign snapshot, but if doesnt exist in campaign any more then use the data in classroom snapshot
+      */
+    buildLevelsData: async ({ commit, rootGetters, dispatch }, campaignHandle, courseInstanceId) => { // campaignHandle necessary
+      let levels = {}
+      await dispatch('campaigns/fetch', campaignHandle, { root: true })
+      const campaignData = rootGetters['campaigns/getCampaignData'](campaignHandle)
+      if (!courseInstanceId) {
+        levels = campaignData.levels
+      } else {
+        try {
+          const courseInstance = await api.courseInstances.get({ courseInstanceID: courseInstanceId })
+          const courseId = courseInstance.courseID
+          const classroomId = courseInstance.classroomID
+
+          // campaign snapshot of the levels
+          const existingCampaignLevels = _.cloneDeep(campaignData.levels)
+
+          // classroom snapshot of the levels for the course
+          const classroom = await api.classrooms.get({ classroomID: classroomId })
+          const classroomCourseLevels = _.find(classroom.courses, { _id: courseId }).levels
+
+          // get levels data for the levels in the classroom snapshot
+          const classroomCourseLevelsData = await api.classrooms.getCourseLevels({ classroomID: classroomId, courseID: courseId })
+
+          const classroomLevelMap = {}
+          for (let level of classroomCourseLevels) {
+            classroomLevelMap[level.original] = level
+          }
+
+          let courseLevelsData = {}
+          for (let level of classroomCourseLevelsData) {
+            let original = level.original
+            if (existingCampaignLevels[original]) {
+              courseLevelsData[original] = existingCampaignLevels[original]
+            } else {
+              // a level which has been removed from the campaign but is saved in the course
+              courseLevelsData[original] = level
+            }
+            // carry over position, nextLevels, first property stored in classroom course, if there are any
+            if (classroomLevelMap[original].position) {
+              courseLevelsData[original].position = classroomLevelMap[original].position
+            }
+            if (classroomLevelMap[original].nextLevels) {
+              courseLevelsData[original].nextLevels = classroomLevelMap[original].nextLevels
+            }
+            if (classroomLevelMap[original].first) {
+              courseLevelsData[original].first = classroomLevelMap[original].first
+            }
+          }
+
+          levels = courseLevelsData
+        } catch (err) {
+          console.error('Error in building levels data', err)
+          // TODO: handle_error_ozaria
+        }
+      }
+      commit('setCurrentLevelsList', levels)
```

Comment:
Currently if there is an error in the courseInstance logic the levels object is still committed.
Is there a chance of any corrupt state being committed?

Reply:
An empty object will be committed in that case because levels has been initialized as {} and is only updated at the end of all the logic that can error out.